Configure timeouts & some improvements

zacblazic opened the pull request:
The main goal of this change is to configure timeouts that work well with the default idle timeout of an upstream load balancer (i.e. [60 seconds](https://docs.aws.amazon.com/elasticloadbalancing/latest/classic/config-idle-timeout.html)).

Additionally brings in some other changes, all of which were inspired by [ingress-nginx](https://github.com/kubernetes/ingress-nginx).

zacblazic committed:
Align client_body_timeout with client_header_timeout

zacblazic committed:
Reduce client_body_buffer_size to 16k

To prevent denial of service attachs, such as http slow attack.

* https://blog.qualys.com/vulnerabilities-research/2011/11/02/how-to-protect-against-slow-http-attacks
* https://www.acunetix.com/blog/web-security-zone/hardening-nginx
* https://www.upguard.com/blog/how-to-build-a-tough-nginx-server-in-15-steps

zacblazic committed:
Disable daemon mode explicitly

* http://nginx.org/en/docs/ngx_core_module.html#daemon

zacblazic committed:
Automatically detect optimal number of worker processes

* http://nginx.org/en/docs/ngx_core_module.html#worker_processes

zacblazic committed:
Use optimal events configuration

* http://nginx.org/en/docs/ngx_core_module.html#multi_accept
* http://nginx.org/en/docs/ngx_core_module.html#worker_connections
* http://nginx.org/en/docs/ngx_core_module.html#use

zacblazic committed:
Enable asynchronous file i/o

* http://nginx.org/en/docs/http/ngx_http_core_module.html#aio
* http://nginx.org/en/docs/http/ngx_http_core_module.html#aio_write

zacblazic committed:
Enable tcp nodelay

* http://nginx.org/en/docs/http/ngx_http_core_module.html#tcp_nodelay

zacblazic committed:
Reduce client timeouts to 60 seconds

* http://nginx.org/en/docs/http/ngx_http_core_module.html#client_body_timeout
* http://nginx.org/en/docs/http/ngx_http_core_module.html#client_header_timeout

zacblazic committed:
Reduce keepalive_timeout to 75 seconds

* http://nginx.org/en/docs/http/ngx_http_core_module.html#keepalive_timeout

zacblazic committed:
Reduce proxy_connect_timeout to 5 seconds

* http://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_connect_timeout

zacblazic committed:
Reduce proxy_read_timeout to 60 seconds

* http://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_read_timeout

zacblazic committed:
Mitigate httpoxy vulnerability

* https://www.nginx.com/blog/mitigating-the-httpoxy-vulnerability-with-nginx/

zacblazic committed:
Move access_log to server block for healthcheck server

zacblazic committed:
Fix indentation for healtcheck server block

zacblazic committed:
Disable keep-alive for healtcheck server

zacblazic committed:
Set worker_shutdown_timeout to 240 seconds

* http://nginx.org/en/docs/ngx_core_module.html#worker_shutdown_timeout

zacblazic commented on config/http.conf:
```diff
@@ -1,18 +1,33 @@
+daemon off;
+
+worker_processes auto;
```
Setting this to `auto` allows nginx to deremine the optimal number of processes based on number of cores.

itskingori replied:
Don't we already do this in `main.conf`? 🤔 

https://github.com/Intellection/docker-nginx-proxy/blob/b87f59bfe036dff06870fd7afb854ebc22aac108/config/main.conf#L1

zacblazic replied:
Yeah, forgot about that. I've moved the configuration to there.

zacblazic commented on config/http.conf:
```diff
@@ -1,18 +1,33 @@
+daemon off;
+
+worker_processes auto;
+worker_shutdown_timeout 240s;
```
I'm currently basing this timeout value on the default [connection draining](https://docs.aws.amazon.com/elasticloadbalancing/latest/classic/config-conn-drain.html) / [deregistration delay](https://docs.aws.amazon.com/elasticloadbalancing/latest/application/load-balancer-target-groups.html#deregistration-delay) of 300 seconds.

itskingori replied:
Noted. We also already do this in `main.conf` ...

https://github.com/Intellection/docker-nginx-proxy/blob/b87f59bfe036dff06870fd7afb854ebc22aac108/config/main.conf#L3

zacblazic replied:
I've moved the configuration to `main.conf`.

zacblazic commented on config/http.conf:
```diff
@@ -1,18 +1,33 @@
+worker_shutdown_timeout 240s;
+
+events {
+  multi_accept        on;
```
Multi-accept and epoll are the optimal settings for Linux >= 2.6 according to the docs.

itskingori replied:
I saw this but left it out because [of this official post](https://www.nginx.com/blog/performance-tuning-tips-tricks/) ...

> We recommend keeping the default value (off), unless you’re sure there’s a benefit to changing it. Start performance testing with the default value to better measure predictable scale.

There's also [a comment there that where someone says](https://www.nginx.com/blog/performance-tuning-tips-tricks/#comment-5043382241) ...

> There's no simple cut-off point beyond which enabling the accept() mutex brings performance benefits. We have used accept_mutex and multi_accept to tune NGINX to give more stable, predictable performance under synthetic benchmarks, but this does not typically apply to production workloads. Some users have applied it in production.
>
> Your mileage will vary. If you're seeing varying latency or high CPU usage on NGINX under high traffic load (new connections), and you cannot scale your hardware, enabling it might make a difference. Leave it disabled unless you measure a clear performance difference, and accept that it rarely brings a clear benefit.

I see that it's on by default in Ingress Nginx Kubernetes and theres' not discussion around why (checked https://github.com/kubernetes/ingress-nginx/issues/2684 and https://github.com/kubernetes/ingress-nginx/pull/2896).

Just to be clear, I'm not opposed to this change. Just sharing some of my finding when I looked into it.

zacblazic replied:
Hmm, I see. We'd need to run some tests to determine the impact, but I thought that since ingress-nginx has it enabled it can't be too bad.

What do you want to do?

itskingori replied:
> Hmm, I see. We'd need to run some tests to determine the impact, but I thought that since ingress-nginx has it enabled it can't be too bad.

Yes. I agree with you here.

> What do you want to do?

We can leave it as is. No objections from me.

zacblazic commented on config/http.conf:
```diff
@@ -1,18 +1,33 @@
   include /etc/nginx/mime.types;
   include /etc/nginx/log.conf;
 
+  aio       threads;
+  aio_write on;
```
Configures asynchronous file I/O which prevents the event loop from blocking.

See: https://www.nginx.com/blog/thread-pools-boost-performance-9x

zacblazic commented on config/http.conf:
```diff
@@ -1,18 +1,33 @@
   server_tokens           off;
   sendfile                on;
   tcp_nopush              on;
+  tcp_nodelay             on;
```
Another performance improvement which can save 200ms for small requests, see [this article](https://www.netguru.com/codestories/nginx-tutorial-performance).

zacblazic commented on config/http.conf:
```diff
@@ -48,6 +63,9 @@ http {
   proxy_set_header    Host                $host;
 
+  # Mitigate httpoxy vulnerability
+  proxy_set_header    Proxy               "";
```
Mitigates [httpoxy vulnerability](https://www.nginx.com/blog/mitigating-the-httpoxy-vulnerability-with-nginx/).

zacblazic commented on config/http.conf:
```diff
@@ -59,11 +77,13 @@ http {
+    listen 18081 default_server;
+
+    access_log off;
+    keepalive_timeout 0;
```
No need for keep-alives here!

itskingori replied:
Very good idea! 👌🏽

zacblazic reviewed (commented):
Some comments...

zacblazic commented on config/http.conf:
```diff
@@ -1,18 +1,33 @@
-  keepalive_timeout       605s;
-  proxy_connect_timeout   60s;
-  proxy_read_timeout      600s;
+  client_body_buffer_size 16k;
+  client_body_timeout     60s;
+  client_header_timeout   60s;
+  keepalive_timeout       75s;
+  proxy_connect_timeout   5s;
+  proxy_read_timeout      60s;
```
All of these values are based on ingress-nginx's defaults.

itskingori replied:
Good start. This was a good suggestion by the way!

itskingori commented:
@zacblazic I think you're missing a `CHANGELOG` entry.

zacblazic committed:
Move events & worker config to main.conf

zacblazic committed:
Update changelog

itskingori reviewed (dismissed):
Great additions! 🎖️ + 🏆

zacblazic committed:
Disable sendfile

Because this only applies when serving static content, which we'll
not be doing directly from the proxy, only from upstream servers.

zacblazic committed:
Enable reset_timedout_connection

Resets timed out connections which avoids keeping an already closed
socket with filled buffers in a closing (FIN_WAIT1) state for a long
time.

* http://nginx.org/en/docs/http/ngx_http_core_module.html#reset_timedout_connection

zacblazic committed:
Separate client/proxy/keepalive sections

zacblazic committed:
Remove setting of client_body_buffer_size

Prefer to let it be the default, we don't really need to have this
set here.

zacblazic committed:
Support websocket connections

zacblazic committed:
Disable port_in_redirect

* http://nginx.org/en/docs/http/ngx_http_core_module.html#port_in_redirect

zacblazic committed:
Prefer to let app.conf set client_max_body_size

zacblazic committed:
Format http.conf

zacblazic committed:
Update changelog

itskingori commented on config/http.conf:
```diff
@@ -51,9 +51,16 @@ http {
   }
 
+  # See https://www.nginx.com/blog/websocket-nginx
+  map $http_upgrade $connection_upgrade {
```
Nit-pick. I'd gotten into the habit of naming 'proxied' variables with a `proxy_` suffix (see configs above) and with the name to match it's use case. Could we do that here too? So that `$http_upgrade` is mapped to `$proxy_connection` ... because it's the now the "connection header on the proxy just mapped out". Makes it easy to know if it was a native or user defined variable at a glance ... and also what it is and where it's used.

```suggestion
  map $http_upgrade $proxy_connection {
```

itskingori commented on config/http.conf:
```diff
@@ -44,9 +49,19 @@ http {
   proxy_http_version  1.1;
-  proxy_set_header    Connection          "";
   proxy_set_header    Host                $host;
+  proxy_set_header    Connection          $connection_upgrade;;
```
If you agree to have to `$proxy_connection` this should change here also. Also note that there are two `;` at the end:

```suggestion
  proxy_set_header    Connection          $proxy_connection;
```

itskingori commented on config/http.conf:
```diff
@@ -51,9 +51,16 @@ http {
 
+  # See https://www.nginx.com/blog/websocket-nginx
+  map $http_upgrade $connection_upgrade {
+    default upgrade;
```
I don't know if this affects anything but it looks like it's `Upgrade` online. Other values are also have the first letter capitalised e.g. `WebSocket` ...

```suggestion
    default Upgrade;
```

itskingori reviewed (commented):
Some comments ... 💬

zacblazic committed:
Rename connection_upgrade to proxy_connection

Co-authored-by: King'ori Maina <j@kingori.co>

itskingori commented on config/http.conf:
```diff
@@ -44,9 +49,19 @@ http {
   proxy_http_version  1.1;
-  proxy_set_header    Connection          "";
   proxy_set_header    Host                $host;
+  proxy_set_header    Connection          $proxy_connection;
+  proxy_set_header    Upgrade             $http_upgrade;
```
Now thinking that maybe we should add these to logging? Thoughts?

zacblazic committed:
Capitalise upgrade

Co-authored-by: King'ori Maina <j@kingori.co>

zacblazic committed:
Add http_upgrade and proxy_connection to log format

itskingori commented on CHANGELOG.md:
```diff
@@ -1,5 +1,28 @@
+* Set `aio_write` to `on`.
+* Set `tcp_nodelay` to `on`.
+* Set `reset_timedout_connection` to `on`.
+* Set `port_in_redirect` to `off`.
```
```suggestion
* Set `port_in_redirect` to `off`.
* Add `http_upgrade` and `proxy_connection` to log format.
```

zacblazic committed:
Update changelog

Co-authored-by: King'ori Maina <j@kingori.co>

itskingori reviewed (dismissed):
Let's give this a spin. 🌪️

itskingori commented on config/log.conf:
```diff
@@ -8,10 +8,12 @@ log_format main_json escape=json
     '"body_bytes_sent":"$body_bytes_sent",'
     '"host":"$host",'
     '"http_connection":"$http_connection",'
+    '"http_upgrade":"$http_upgrade",'
```
Oops. Not alphabetical. 😅 Nit-pick.

zacblazic committed:
Sort log format fields alphabetically

zacblazic committed:
Set command in dockerfile explicitly

Overwrites the command in the base image which specifies "daemon off".
Ensure we're not passing "daemon off" twice, which resultes in an error.
